### ISSUE TITLE
Question / Confirmation on arange in encode_text

### ISSUE THREAD

**Issue by AwePhD** [openai/CLIP#371]
Hello,

This is a minor question about the code, I want to be sure I do not let slip any subtleties.

In L354 of `model.py`, there is the final step to extract the text

```python
# x.shape = [batch_size, n_ctx, transformer.width]
x = x[torch.arange(x.shape[0]), text.argmax(dim=-1)] @ self.text_projection
```

- `self.text_projection`  is the last projection to have the text features
- `text.argmax(dim=-1)` picks the features of the EOT token. 

Why there is a `torch.arange(x.shape[0])`? It could be `x[:, text.argmax(dim=-1)]`, right?

Thanks for the work, code and model.

**Comment by jongwook** [comment]
We need to extract the features (of shape `transformer.width`) from specific different locations of the EOT token across the batch, so the shape becomes `[batch_size, transformer.width]`.

If we did `x[:, text.argmax(dim=-1)]`, it simply selects the features at any of the argmax indices across the batch, so the shape would become `[batch_size, batch_size, transformer.width]`. 

It could be simpler to use `index_select` instead and still achieve the same effect, but I found myself fewer mistakes by explicitly indexing with `torch.arange()`.
